src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -0,0 +1,171 @@
+
+    server_addr = params->sockaddr->addr;
+    status = ucs_socket_create(server_addr->sa_family, SOCK_STREAM, &cep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the fd to non-blocking mode. (so that connect() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(cep->fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    /* try to connect to the server */
+    status = ucs_socket_connect(cep->fd, server_addr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        goto err_close_socket;
+    }
+    ucs_assert ((status == UCS_OK) || (status == UCS_INPROGRESS));
+
+    if (status == UCS_OK) {
+        cep->state &= ~UCT_TCP_SOCKCM_EP_INIT;
+        cep->state |= UCT_TCP_SOCKCM_EP_CONNECTED;
+        /* TODO: start sending the user's data */
+
+        goto out_print;
+    }
+
+    sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "client sa_arg_ctx");
+    if (sa_arg_ctx == NULL) {
+        ucs_error("failed to allocate memory for a client_ctx");
+        status = UCS_ERR_NO_MEMORY;
+        goto err_close_socket;
+    }
+
+    sa_arg_ctx->fd = cep->fd;
+    sa_arg_ctx->ep = cep;
+
+    /* Adding the arg to a list on the cm for cleanup purposes */
+    ucs_list_add_tail(&tcp_sockcm->sa_arg_list, &sa_arg_ctx->list);
+
+    events    = UCS_EVENT_SET_EVWRITE;    /* wait until connect() completes */
+    async_ctx = tcp_sockcm->super.iface.worker->async;
+    status    = ucs_async_set_event_handler(async_ctx->mode, cep->fd, events,
+                                            uct_tcp_sa_data_handler, sa_arg_ctx,
+                                            async_ctx);
+    if (status != UCS_OK) {
+        goto err_close_socket;
+    }
+
+out_print:
+    ucs_debug("created a TCP SOCKCM endpoint (fd=%d) on tcp cm %p, "
+              "remote addr: %s", cep->fd, tcp_sockcm,
+              ucs_sockaddr_str(server_addr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
+
+    return status;
+
+err_close_socket:
+    close(cep->fd);
+err:
+    return status;
+}
+
+UCS_CLASS_INIT_FUNC(uct_tcp_sockcm_ep_t, const uct_ep_params_t *params)
+{
+    ucs_status_t status;
+
+    status = uct_cm_check_ep_params(params);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &params->cm->iface);
+
+    self->wireup.priv_pack_cb = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                                params->sockaddr_pack_cb : NULL;
+    self->disconnect_cb       = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_SOCKADDR_DISCONNECT_CB) ?
+                                params->disconnect_cb : NULL;
+    self->user_data           = (params->field_mask &
+                                 UCT_EP_PARAM_FIELD_USER_DATA) ?
+                                params->user_data : NULL;
```

Comment:
can we move this code + uct_cm_check_ep_params + connect_cb to new `uct_base_cm_ep_t` class? and the same for rdmacm